Introduce cmake variable USE_SYSTEM_ROCM

trixirt opened the pull request:
Fedora has native system ROCm support.  So the various ROCm packages can be treated as normal packages and do not require special handling.

So introduce the cmake variable USE_SYSTEM_ROCM and turn it off by default.  Then add minimal logic in LoadHIP to use it.

Fixes #ISSUE_NUMBER


cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

jithunnair-amd commented on cmake/public/LoadHIP.cmake:
```diff
@@ -1,246 +1,273 @@
+  endif()
+endif()
 
+if (PYTORCH_FOUND_HIP)
```
Is `PYTORCH_FOUND_HIP` set for the system ROCm case?

jithunnair-amd commented on cmake/public/LoadHIP.cmake:
```diff
@@ -1,246 +1,273 @@
+  math(EXPR TORCH_HIP_VERSION "(${HIP_VERSION_MAJOR} * 100) + ${HIP_VERSION_MINOR}")
+  math(EXPR ROCM_VERSION_DEV_INT "(${HIP_VERSION_MAJOR} * 1000) + ${HIP_VERSION_MINOR} * 100")
+  
+  find_package(hsa-runtime64 REQUIRED)
```
Instead of introducing a duplicated list of packages (which will increase maintenance burden), can we use the `foreach` CMake command to loop through a package list, and use the same package list for both system and non-system ROCm cases?

jithunnair-amd reviewed (commented):
Which version of ROCm is currently provided in Fedora as the system ROCm? Also, how would this newly introduced code be tested? My concern is that we will add this "feature", but without any regular testing, it will not stay up-to-date.

trixirt commented:
> Which version of ROCm is currently provided in Fedora as the system ROCm? Also, how would this newly introduced code be tested? My concern is that we will add this "feature", but without any regular testing, it will not stay up-to-date.

I am the Fedora pytorch maintainer as well as the maintainer of most of these rocm packages
Here is the in-process changes for 5.7.1[
](https://src.fedoraproject.org/rpms/python-torch/c/ba89707404cb66de4862e6262f36ba4f5e83985d?branch=rawhide)
This will be updated to 6.0 soon.

Commit:
Introduce cmake variable USE_SYSTEM_ROCM

Fedora has native system ROCm support.  So the various ROCm
packages can be treated as normal packages and do not require
special handling.

So introduce the cmake variable USE_SYSTEM_ROCM and turn it
off by default.  Then add minimal logic in LoadHIP to use it.

Signed-off-by: Tom Rix <trix@redhat.com>

trixirt commented:
> > > Which version of ROCm is currently provided in Fedora as the system ROCm? Also, how would this newly introduced code be tested? My concern is that we will add this "feature", but without any regular testing, it will not stay up-to-date.
> > 
> > 
> > I am the Fedora pytorch maintainer as well as the maintainer of most of these rocm packages Here is the in-process changes for 5.7.1[ rocm changes for pytorch ](https://src.fedoraproject.org/rpms/python-torch/c/ba89707404cb66de4862e6262f36ba4f5e83985d?branch=rawhide) This will be updated to 6.0 soon.
> 
> Would it make sense to create a separate cmake file for Fedora, say `LoadFedoraSystemRocm.cmake` and include it instead of `LoadHIP.cmake` based on the new flag. Another approach to consider would be to use `if() return() endif()` construct if making changes to `LoadHIP.cmake` to avoid giant `if() else()` blocks making it difficult to read.
> 
> Also, the `USE_SYSTEM_ROCM` name seems too generic. If this is only for `Fedora` like distros then maybe it should be renamed to `USE_FEDORA_LIKE_SYSTEM_ROCM` so users don't set it in other OS environment and expect it to work. And, maybe add a check to ensure that the OS is Fedora like, test `ID_LIKE`, when the flag is set by a user and emit an error if flag defined incorrectly on non-Fedora OS? Aside: Is there a way to have both Fedora system ROCm and ROCm under `/opt` installed? Any conflicts with package names?

Fedora packaging requires using the system location.  This allows rocm packages to be integrated into other packages without those packages having to know the special install locations and environment variables.

trixirt commented:
This change is more invasive. it limits the USE_SYSTEM_ROCM logic to requiring finding ROCM and HIP packages in the standard location and not setting up variables to directories that do not exist.

Some issues
# roctx is part of roctracer                                                                                                    
find_library(ROCM_ROCTX_LIB roctx64 HINTS ${ROCM_PATH}/lib) 
roctracer is never find_package-ed.  And I had to comment out this line in my test.

if(ROCM_VERSION_DEV VERSION_GREATER_EQUAL "5.7.0")                                                                              
    # check whether hipblaslt is using its own datatype  
should check the version of hipblaslt instead of doing a test compile.
This compile fails for me and I had to comment it out.

Mystro256 commented:
> Fedora packaging requires using the system location

Debian also has this requirement as far as I know. @cgmb should be able to comment.

srinivamd commented:
> This change is more invasive. it limits the USE_SYSTEM_ROCM logic to requiring finding ROCM and HIP packages in the standard location and not setting up variables to directories that do not exist.
> 
> Some issues roctx is part of roctracer find_library(ROCM_ROCTX_LIB roctx64 HINTS ${ROCM_PATH}/lib) roctracer is never find_package-ed. And I had to comment out this line in my test.
> 
> if(ROCM_VERSION_DEV VERSION_GREATER_EQUAL "5.7.0") check whether hipblaslt is using its own datatype should check the version of hipblaslt instead of doing a test compile. This compile fails for me and I had to comment it out.

Since the change is invasive, I think this should be in a separate cmake file and not part of `LoadHIP.cmake`  so changes, bugs etc are confined to a module and the two can independently evolve. Could you please update the PR by separating out the cmake changes?

srinivamd commented:
> > > > Which version of ROCm is currently provided in Fedora as the system ROCm? Also, how would this newly introduced code be tested? My concern is that we will add this "feature", but without any regular testing, it will not stay up-to-date.
> > > 
> > > 
> > > I am the Fedora pytorch maintainer as well as the maintainer of most of these rocm packages Here is the in-process changes for 5.7.1[ rocm changes for pytorch ](https://src.fedoraproject.org/rpms/python-torch/c/ba89707404cb66de4862e6262f36ba4f5e83985d?branch=rawhide) This will be updated to 6.0 soon.
> > 
> > 
> > Would it make sense to create a separate cmake file for Fedora, say `LoadFedoraSystemRocm.cmake` and include it instead of `LoadHIP.cmake` based on the new flag. Another approach to consider would be to use `if() return() endif()` construct if making changes to `LoadHIP.cmake` to avoid giant `if() else()` blocks making it difficult to read.
> > Also, the `USE_SYSTEM_ROCM` name seems too generic. If this is only for `Fedora` like distros then maybe it should be renamed to `USE_FEDORA_LIKE_SYSTEM_ROCM` so users don't set it in other OS environment and expect it to work. And, maybe add a check to ensure that the OS is Fedora like, test `ID_LIKE`, when the flag is set by a user and emit an error if flag defined incorrectly on non-Fedora OS? Aside: Is there a way to have both Fedora system ROCm and ROCm under `/opt` installed? Any conflicts with package names?
> 
> Fedora packaging requires using the system location. This allows rocm packages to be integrated into other packages without those packages having to know the special install locations and environment variables.

Not sure the issues raised were adequately addressed or what "special" "standard" locations mean. FWIW, here https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.html

cgmb commented on cmake/public/LoadHIP.cmake:
```diff
@@ -1,154 +1,168 @@
+  set(PYTORCH_FOUND_HIP TRUE)
+  set(FOUND_ROCM_VERSION_H FALSE)
 
+  find_package(ROCM CONFIG REQUIRED)
```
This doesn't seem to be used anywhere. `find_package(ROCM CONFIG REQUIRED)` searches for rocm-cmake. Note the [README.md](https://github.com/ROCm/rocm-cmake/tree/rocm-6.0.0).

cgmb commented:
@cdluminate is the package maintainer for PyTorch on Debian and would probably be interested in this change.

> Also, the `USE_SYSTEM_ROCM` name seems too generic. If this is only for `Fedora` like distros then maybe it should be renamed to `USE_FEDORA_LIKE_SYSTEM_ROCM` so users don't set it in other OS environment and expect it to work.

Standard `find_package` calls should also 'just work' with Debian, Ubuntu and Gentoo system packages.

GZGavinZhao commented:
Hi, I'm the maintainer of ROCm for Solus. I'd like to report that currently the CMake build process works for us out of the box i.e. all the `find_package` calls can detect packages without patching. (see the package recipe file [here](https://github.com/getsolus/packages/blob/main/packages/py/pytorch/package.yml)). This is with PyTorch 2.1.1 + ROCm 5.5. Does Fedora have the [`rocm-core`](https://github.com/ROCm/rocm-core) package in its repositroy? We also had to patch the CMake files until we realized we need `rocm-core` for HIP to be correctly detected.

cdluminate commented:
There are still some not-yet-upstreamed modifications on the Debian side: https://salsa.debian.org/deeplearning-team/pytorch/-/blob/master/debian/patches/dirtyhack.patch?ref_type=heads
I'll split the patch and submit a part of them soon.

cdluminate commented:
Debian's ROCm is still a work in progress. Some of the pytorch-rocm dependencies are still missing. So I won't be able to test `USE_SYSTEM_ROCM` soon. But I'm sure I'll test it when it is time to prepare pytorch-rocm on debian side.

cgmb commented:
> Does Fedora have the [`rocm-core`](https://github.com/ROCm/rocm-core) package in its repositroy? We also had to patch the CMake files until we realized we need `rocm-core` for HIP to be correctly detected.

I have never seen a single example of the rocm-core library being used correctly. I would strongly oppose packaging it for Debian. PyTorch should be able to build without it.

GZGavinZhao commented:
> > Does Fedora have the [`rocm-core`](https://github.com/ROCm/rocm-core) package in its repositroy? We also had to patch the CMake files until we realized we need `rocm-core` for HIP to be correctly detected.
> 
> I have never seen a single example of the rocm-core library being used correctly. I would strongly oppose packaging it for Debian. PyTorch should be able to build without it.

I agree, but currently I can't build PyTorch without it, mainly for 2 reasons:
1. [`LoadHIP.cmake`](https://github.com/pytorch/pytorch/blob/de4d48df34c0e8bcd03d96bcdfd12e263e9889ff/cmake/public/LoadHIP.cmake#L52) relies on parsing `rocm_version.h` to detect current ROCm version. 
2. The [hipify tool](https://github.com/pytorch/pytorch/blob/de4d48df34c0e8bcd03d96bcdfd12e263e9889ff/torch/utils/hipify/cuda_to_hip_mappings.py#L42) relies on parsing `rocm_version.h` to determine whether the include paths needs to be backward compatible with non-FHS installs, i.e. whether to use the old `rocblas.h` or the new `rocblas/rocblas.h`.

cgmb commented:
>     1. [`LoadHIP.cmake`](https://github.com/pytorch/pytorch/blob/de4d48df34c0e8bcd03d96bcdfd12e263e9889ff/cmake/public/LoadHIP.cmake#L52) relies on parsing `rocm_version.h` to detect current ROCm version.

This is a good example of what I was talking about. If you look at what the rocm-core version is ultimately used for, it's code like this:

```cmake
if(ROCM_VERSION_DEV VERSION_GREATER_EQUAL "5.7.0")
  find_package_and_print_version(hipblaslt REQUIRED)
endif()
```

This logic is needlessly brittle. The build needs to know "is hipblaslt available?" but rather than directly asking that question, checks the rocm-core version number and reasons that hipblaslt must be available if the version number is 5.7 or greater. That reasoning comes with a lot of assumptions that may be wrong.

Robust code should prefer feature detection to version checking. If you want to know if you can use hipblaslt, do `find_package(hipblaslt)` and check `hipblaslt_FOUND`. Use that to set the state of a `WITH_HIPBLASLT` cache variable that you can use later on in your build code. If the user manually set `-DWITH_HIPBLASLT=ON` and hipblaslt was not found, then emit an error.

Here's another place where it's used:

```cmake
if(ROCM_VERSION_DEV VERSION_GREATER_EQUAL "6.0.0")
  list(APPEND HIP_CXX_FLAGS -DHIPBLAS_V2)
endif()
```

This is another example of a misuse of the rocm-core library. This should just be checking the hipblas version, rather than checking the rocm-core version and using that to reason about hipblas. The find_package call for a library will provide version information for each library.

```cmake
if(hipblas_VERSION VERSION_GREATER_EQUAL "2.0.0")
  list(APPEND HIP_CXX_FLAGS -DHIPBLAS_V2)
endif()
```

It also defines a ROCM_VERSION for use in the C++ code, which results in more misuse, like this:

```cpp
// hipSparse Generic API ROCm 5.2
#if ROCM_VERSION >= 50200
#define AT_USE_HIPSPARSE_GENERIC_52_API() 1
#else
#define AT_USE_HIPSPARSE_GENERIC_52_API() 0
#endif
```

which checks the rocm-core version to reason about the hipSPARSE API rather than directly using the hipSPARSE version, like so:
```cpp
// hipSparse Generic API ROCm 5.2
#if (hipsparseVersionMajor > 2) || (hipsparseVersionMajor == 2) && (hipsparseVersionMinor >= 1)
#define AT_USE_HIPSPARSE_GENERIC_52_API() 1
#else
#define AT_USE_HIPSPARSE_GENERIC_52_API() 0
#endif
```

>     2. The [hipify tool](https://github.com/pytorch/pytorch/blob/de4d48df34c0e8bcd03d96bcdfd12e263e9889ff/torch/utils/hipify/cuda_to_hip_mappings.py#L42) relies on parsing `rocm_version.h` to determine whether the include paths needs to be backward compatible with non-FHS installs, i.e. whether to use the old `rocblas.h` or the new `rocblas/rocblas.h`. Therefore, if your system has a FHS ROCm with backward compatible reorg turned off, which I assume is what most distributions have, without `rocm-core` you will have wrong header file locations.

It should directly test for include style using `__has_include`. The rocm-core version is not guaranteed to tell you which style was used, since as you mention, that depends on what build options were set when the libraries were compiled.

And, consider the current state of Debian Unstable (which is likely to be what ships in Ubuntu 24.04 LTS in a few months). It contains HIP from ROCm 5.2 and rocBLAS from ROCm 5.5. What value should be returned from rocm-core? The version returned from rocm-core is useless.

As far as I know, there is no correct way to use the rocm-core library. It should never have been created, as it encourages developers to make brittle assumptions about their environment, rather than directly checking for the properties they care about.

trixirt commented:
This change set will need to be rebased for ROCm 6.0, which I am in the middle of getting that last couple of packages into Fedora.

My interpretation of USE_SYSTEM_ comes from the top level CMakeLists.txt on how some of the third_party/ is build or reused from the os.  In the case of Fedora these come from Fedora, I packaged the few up that were not already in Fedora.  For our ROCm packages, they are all treated like other packages with their prefix being /usr, not /opt/rocm and there should be no necessity to set ROCM_PATH or any other environmental variable.

The dependence of the current configury on these enviromental variables and the implicit /opt/rocm install location is why I am doing this change.  As pointed out above there are other brittle configury issues.  Mostly I want to get to a minimal set of find_packages() and remove the fiddly bits.  I problem I have with having another cmake file file is the logic of the fiddly bits, it is brittle and magic and would need constant attention.

It is not critical we solve the problem _right now_ but all the necessary ROCm packages will be in no later than Fedora 41 and general users will stumble over this.

GZGavinZhao commented:
@trixirt Sorry if I wasn't being clear earlier. I don't think it is necessary to introduce `USE_SYSTEM_ROCM`. I maintain ROCm for Solus and we didn't patch anything when building PyTorch, only defined some environment variables, yet `LoadHIP.cmake` still managed to find every library. If it didn't worked for Fedora OOTB, my biggest suspicion is that  `rocm-core` is not being installed. As @cgmb pointed out, the need for `rocm-core` isn't really ideal, but that would be a separate patch that we could submit later. If you really don't want to install `rocm-core`, you could also probably patch CMake to hardcode the ROCm version instead of looking for `rocm_version.h`.

Feel free to reference our [package recipes](https://github.com/getsolus/packages) if you'd like to compare and contrast.
